File: arrays/medium/rearrange.cpp
#include <bits/stdc++.h>
using namespace std;


// variety 1 : pos == neg
// vector<int> rearrange(vector<int> &arr){
//     vector<int> result;
//     int n = arr.size();
//     int posIndex = 0; int negIndex = 1;
//     for(int i = 0; i < n; i++){
//         if(arr[i] < 0){
//             result[negIndex] = arr[i];
//             negIndex += 2;
//         } else{
//             result[posIndex] = arr[i];
//             posIndex += 2;
//         }
//     }
//     return result;
// }

vector<int> rearrange(vector<int> &arr){
    vector<int> pos, neg;
    int n = arr.size();
    for(int i = 0; i < n; i++){
        if(arr[i] > 0){
            pos.push_back(arr[i]);
        } else{
            neg.push_back(arr[i]);
        }
    }
    if(neg.size() < pos.size()){
        for(int i = 0; i < neg.size(); i++){
            arr[2* i] = pos[i];
            arr[2*i + 1] = neg[i];
        }
        int index = neg.size() * 2;
        for(int i = neg.size(); i < pos.size(); i++){
            arr[index] = pos[i];
            index++;
        }
    } else{
        for(int i = 0; i < pos.size(); i++){
            arr[2* i] = pos[i];
            arr[2*i + 1] = neg[i];
        }
        int index = pos.size() * 2;
        for(int i = pos.size(); i < neg.size(); i++){
            arr[index] = neg[i];
            index++;
        }
    }
    return arr;
}
int main(){
    vector<int> someArr = {3, 1, -2, -2, 2, -4, 3, 4, 1, 4, 2, -2};
    vector<int> result = rearrange(someArr);
    for(auto el: result){
        cout << el << " ";
    }
    return 0;
}

File: arrays/medium/rearrange.py
def rearrangeBrute(arr):
    pos, neg = [], []
    for i in range(len(arr)):
        if arr[i] > 0:
            pos.append(arr[i])
        else:
            neg.append(arr[i])
    for i in range(len(pos)):
        arr[i * 2] = pos[i]
    for i in range(len(neg)):
        arr[i * 2 + 1] = neg[i]
    return arr

arr = [1, -2, 3, -4, 5, -6]
resultBrute = rearrangeBrute(arr)
print(resultBrute)

def rearrangeOptimal(arr):
    pos_inx = 0
    neg_index = 1
    result =[0] * len(arr)
    i = 0
    while i < len(arr):
        if arr[i] > 0: 
            result[pos_inx] = arr[i]
            pos_inx += 2
            i+= 1
        else:
            result[neg_index] = arr[i]
            neg_index += 2
            i += 1
    return result

resultOptimal = rearrangeOptimal([1, -2, 3, -4, 5, -6])
print(resultOptimal)


def v2Rearrange(arr):
    pos = []
    neg = []
    i = 0
    for i in range(len(arr)):
        if arr[i] > 0:
            pos.append(arr[i])
        else:
            neg.append(arr[i])
    print(f'Positive: {pos}, Negative: {neg}')
    min_len = min(len(pos), len(neg))
    for i in range(min_len):
        arr[i * 2] = pos[i]
    for i in range(min_len):
        arr[i * 2 + 1] = neg[i]
    print(arr)
    if len(pos) > len(neg):
        arr[min_len * 2:] = pos[min_len:]
    else:
        arr[min_len * 2:] = neg[min_len:]
    return arr

resultOptimal = v2Rearrange([1, 2, 3, -4, 5, -6])
print(resultOptimal)     